### PR TITLE
CP-49910: mv `scripts/link-vms-by-sr.py` `python3/libexec`: pyright, pylint

### PR DESCRIPTION
**Pull request by ashwin9390**
Main ticket CP-49100 :Move python3 scripts from scripts directory to python3 directory.

Sub task CP-49910:

- Modified Makefile in python3 directory  to include link-vms-by-sr.py
- Removed link-vms-by-sr.py from scripts directory
- Removed link-vms-by-sr.py from scripts/Makefile
- Fixed pylint issue

 Signed-off-by: Ashwinh <ashwin.h@cloud.com>

### CODE REVIEW

File: python3/libexec/link-vms-by-sr.py [hunk 1]
```diff
@@ -2,6 +2,10 @@
 # Populate a directory of symlinks partitioning VMs by SR
 # (c) Anil Madhavapeddy, Citrix Systems Inc, 2008
 
+# pyright: reportFunctionMemberAccess=false
+# pyright: reportOptionalMemberAccess=false, reportAttributeAccessIssue=false
+# pyright: reportPossiblyUnboundVariable=false
```
Comment (psafont) [inline python3/libexec/link-vms-by-sr.py:7]: Please address the warnings instead of ignoring them

Comment (ashwin9390) [reply python3/libexec/link-vms-by-sr.py:7]: Okay

Comment (ashwin9390) [reply python3/libexec/link-vms-by-sr.py:7]: The original pyright erros were :home/runner/work/xen-api/xen-api/python3/libexec/link-vms-by-sr.py
  /home/runner/work/xen-api/xen-api/python3/libexec/link-vms-by-sr.py:71:20 - error: "login_with_password" is not a known attribute of "None" (reportOptionalMemberAccess)
  /home/runner/work/xen-api/xen-api/python3/libexec/link-vms-by-sr.py:71:20 - error: Cannot access attribute "login_with_password" for class "object"
    Attribute "login_with_password" is unknown (reportAttributeAccessIssue)
  /home/runner/work/xen-api/xen-api/python3/libexec/link-vms-by-sr.py:71:20 - error: Cannot access attribute "login_with_password" for class "function"
    Attribute "login_with_password" is unknown (reportFunctionMemberAccess)
  /home/runner/work/xen-api/xen-api/python3/libexec/link-vms-by-sr.py:94:63 - error: "src" is possibly unbound (reportPossiblyUnboundVariable)
  /home/runner/work/xen-api/xen-api/python3/libexec/link-vms-by-sr.py:94:68 - error: "targ" is possibly unbound (reportPossiblyUnboundVariable)
  /home/runner/work/xen-api/xen-api/python3/libexec/link-vms-by-sr.py:96:20 - error: "logout" is not a known attribute of "None" (reportOptionalMemberAccess)
  /home/runner/work/xen-api/xen-api/python3/libexec/link-vms-by-sr.py:96:20 - error: Cannot access attribute "logout" for class "object"
    Attribute "logout" is unknown (reportAttributeAccessIssue)
  /home/runner/work/xen-api/xen-api/python3/libexec/link-vms-by-sr.py:96:20 - error: Cannot access attribute "logout" for class "function"
    Attribute "logout" is unknown (reportFunctionMemberAccess)

Above  errors are seen when performing static code check , login_with_password is part of XenAPI module 

When the commit is pushed .pyright fails with these error  message when running test python3.11 to overcome this issue ,i have used  pyright linter to ignore above error : reportFunctionMemberAccess=false
# pyright: reportOptionalMemberAccess=false, reportAttributeAccessIssue=false
# pyright: reportPossiblyUnboundVariable=false

Comment (psafont) [reply python3/libexec/link-vms-by-sr.py:7]: They look like legitimate issues to me, and therefore mustn't be ignored

Comment (ashwin9390) [reply python3/libexec/link-vms-by-sr.py:7]: Acknowledged.

**Review by bernhardkaindl — APPROVED** [review]
Updated @ashwin9390's commit and added 3 commits:
1. Updated to remove globally ignoring the `pyright` warnings that @pauldu asked to fix and updated the Makefile diff to not conflict with 5722 and by ready for alphabetic ordering of installing files in the Makefile to reduce conflicts when adding files and make it easy to find files because they can be looked up alphabetically.
2. Fix `pyright`: Define the `stubPath` to make `pyright` find the typing stub `python3/stubs/XenAPI.pyi`, which fixes most `pyright` warnings. This is also effective when using `VSCode` as `pylance` is based on `pyright`. Fix the warning on `src` and `targ` not set by moving them out of the `try:` clause: Setting them is not expected to cause an exception because `"{}".format(var)` is expected to always work, even if `var` would be `None`, or even a `function` object.
3. `pylint/missing-docstrings`: Add missing `docstrings` and comments which describe the script, functions and sections in `main()`.
4. `pylint/bare-except`: Replace two `except:` with `except: OSError` as `mkdir` and `symlink` are not expected to raise exceptions other than those that are subclasses of `OSError` (the expectation is that they only raise `OSError` itself).

On commit 4 (fix `pylint:bare-except`):
- Fixing `pylint:bare-except` is optional for us because we disabled it in `pyproject.toml` to make the Python3 migration easier.
- But bare except is still a statement that should be avoided, as explained at https://pylint.pycqa.org/en/latest/user_guide/messages/warning/bare-except.html:
  > *A bare ``except:`` clause will catch ``SystemExit`` and ``KeyboardInterrupt`` exceptions, making it harder to interrupt a program with ``Control-C``, and can disguise other problems. If you want to catch all exceptions that signal program errors, use ``except Exception:`` (bare except is equivalent to ``except BaseException:``).*
- However, to be sure that the change to fix this warning is correct, it needs to be very carefully reviewed and tested. As we did not have the capacity to do either because of other priorities, it was correct to instead disable this warning. But because now, we no longer have such urgencies and should not have problems to carefully review and test moved code, and the main part of the Python3 migration is done, we can fix such warnings as well.
- In this case, it should be clear that `mkdir` and `symlink` can only `raise`  exceptions of type `OSError` (or subclasses of it), and it would be quite safe to fix these two cases. I also tested at least the common case of exception `mkdir`, the already existing directory with this change:
```py
/root/link-vms-by-sr.py -d  /tmp/bk/foo
Directory /tmp/bk/foo/by-sr/65308c16-ae9c-4f07-100b-5f3b6fbf5ce2 already exists, skipping
Directory /tmp/bk/foo/by-sr/03086bb3-efa1-a74e-fbc5-16077bf518e7 already exists, skipping
```
I'd not recommend requiring to fix all bare `except:` cases yet:
- unless we are able to spend the time to fix all cases of `except:` with careful review and testing, it would be safer to keep those as-is.

**Review by bernhardkaindl — APPROVED** [review]
I updated the 2nd commit to now pass CI with the other scripts:

- Because the typing stubs for XenAPI.py is now used also by `pyright`, it recognizes and warns when `xenapi.login_with_password()` is called without the (optional) version and originator arguments that are now common practise in the current code: 
  - While `xenapi.login_with_password()` still works without passing the version and the originator, adding them fixes the warnings and I think we should use the new arguments also in `hfx_filename`.
 
- To fix a few other places, the typing stub `python3/stubs/xcp/logger.pyi` now also needs to declare the functions `info()` and `critical()`.
- Along with setting `stubPath` in the `pyright`'s section `[tool.pyright]` in `pyproject.toml`, we also need to set `reportMissingModuleSource = false` to silence the `reportMissingModuleSource` warning.
  - Otherwise, we'd have to make the source of `xcp.logger` source available for the `pyright` check, which we don't have currently (would need a `git clone` that is bad, or we'd have to make a `xenserver-xcp` package or so) and not really need it.

**Review by psafont — APPROVED** [review]

**Pull request merged**